Mathlib/GroupTheory/SpecificGroups/Coxeter.lean

```diff
@@ -0,0 +1,456 @@
+
+@[simp]
+theorem ofCoxeterGroup_apply {X : Type*} (D : Matrix X X ℕ) (x : X) :
+    CoxeterSystem.ofCoxeterGroup X D x = CoxeterGroup.of D x :=
+  rfl
+
+theorem map_relations_eq_reindex_relations (e : B ≃ B') :
+    (MulEquiv.toMonoidHom (FreeGroup.freeGroupCongr e)) '' CoxeterGroup.Relations.toSet M =
+    CoxeterGroup.Relations.toSet (reindex e e M) := by
+  simp [CoxeterGroup.Relations.toSet, CoxeterGroup.Relations.ofMatrix]
+  apply le_antisymm
+  · rw [Set.le_iff_subset]; intro _
+    simp only [Set.mem_image, Set.mem_range, Prod.exists, Function.uncurry_apply_pair,
+      forall_exists_index, and_imp]
+    intro _ hb b _ heq; rw [← heq]
+    use (e hb); use (e b); aesop
+  · rw [Set.le_iff_subset]; intro hb'
+    simp only [Set.mem_range, Prod.exists, Function.uncurry_apply_pair, Set.mem_image,
+      forall_exists_index]
+    intro b1' b2' heq; rw [← heq]
+    use ((FreeGroup.freeGroupCongr e).symm hb')
+    exact ⟨by use (e.symm b1'); use (e.symm b2'); aesop, by aesop⟩
+
+/-- Coxeter groups of isomorphic types are isomorphic. -/
+def equivCoxeterGroup (e : B ≃ B') : CoxeterGroup M ≃* CoxeterGroup (reindex e e M) :=
+  QuotientGroup.congr (Subgroup.normalClosure (CoxeterGroup.Relations.toSet M))
+    (Subgroup.normalClosure (CoxeterGroup.Relations.toSet (reindex e e M)))
+    (FreeGroup.freeGroupCongr e) (by
+      have := Subgroup.map_normalClosure (CoxeterGroup.Relations.toSet M)
+        (FreeGroup.freeGroupCongr e).toMonoidHom (FreeGroup.freeGroupCongr e).surjective
+      rwa [map_relations_eq_reindex_relations] at this)
+
+theorem equivCoxeterGroup_apply_of (b : B) (M : Matrix B B ℕ) (e : B ≃ B') :
+    (equivCoxeterGroup e) (CoxeterGroup.of M b) = CoxeterGroup.of (reindex e e M) (e b) :=
+  rfl
+
+theorem equivCoxeterGroup_symm_apply_of (b' : B') (M : Matrix B B ℕ) (e : B ≃ B') :
+    (equivCoxeterGroup e).symm (CoxeterGroup.of (reindex e e M) b') =
+    CoxeterGroup.of M (e.symm b') :=
+  rfl
+
+/-- Reindex a Coxeter system through a bijection of the indexing sets. -/
+@[simps]
+protected def reindex (cs : CoxeterSystem M W) (e : B ≃ B') :
+    CoxeterSystem (reindex e e M) W :=
+  ofMulEquiv (cs.mulEquiv.trans (equivCoxeterGroup e))
+
+@[simp]
+theorem reindex_apply (cs : CoxeterSystem M W) (e : B ≃ B') (b' : B') :
+    cs.reindex e b' = cs (e.symm b') :=
+  rfl
+
+/-- Pushing a Coxeter system through a group isomorphism. -/
+@[simps]
+protected def map (cs : CoxeterSystem M W) (e : W ≃* H) : CoxeterSystem M H :=
+  ofMulEquiv (e.symm.trans cs.mulEquiv)
+
+@[simp]
+theorem map_apply (cs : CoxeterSystem M W) (e : W ≃* H) (b : B) : cs.map e b = e (cs b) :=
+  rfl
+
+end CoxeterSystem
+
+namespace CoxeterMatrix
+
+open Matrix
+
+variable (n : ℕ)
+
+/-- The Coxeter matrix of family A(n).
+
+The corresponding Coxeter-Dynkin diagram is:
+```
+    o --- o --- o ⬝ ⬝ ⬝ ⬝ o --- o
+```
+-/
+abbrev Aₙ : Matrix (Fin n) (Fin n) ℕ :=
+  Matrix.of fun i j : Fin n =>
+    if i = j then 1
+      else (if i = n - 1 ∨ j = n - 1 then 2 else 3)
```

Comment:
Could you double check the definition of all the matrices, because it seems to me that some of these are wrong. For example, for $A_n$, the coefficient $(i,j)$ for i different from $j$ is either $3$ when $j = i\pm 1$ and $2$ otherwise?

Reply:
I updated the definitions for the infinite families.  I *think* they are correct now but please take another look if you can.